src/lua/fiber.c

```diff
@@ -802,20 +802,23 @@ static int
 lbox_fiber_join(struct lua_State *L)
 {
 	struct fiber *fiber = lbox_checkfiber(L, 1);
+	double timeout = TIMEOUT_INFINITY;
 	struct lua_State *child_L = fiber->storage.lua.stack;
 	struct error *e = NULL;
 	int num_ret = 0;
 	int coro_ref = 0;
 
 	if (!(fiber->flags & FIBER_IS_JOINABLE))
 		luaL_error(L, "the fiber is not joinable");
-	fiber_join(fiber);
+	if (lua_isnumber(L, 2))
+		timeout = lua_tonumber(L, 2);
+	int rc = fiber_join_timeout(fiber, timeout);
 
 	if (child_L != NULL) {
 		coro_ref = lua_tointeger(child_L, -1);
 		lua_pop(child_L, 1);
 	}
-	if (fiber->f_ret != 0) {
+	if (fiber->f_ret != 0 || rc != 0) {
```

Comment:
AFAIU you don't need to check f_ret anymore: rc is set to f_ret or to -1 on timeout.

Reply:
fixed